core/envelope-sim/src/main/java/fr/sncf/osrd/envelope/part/EnvelopePart.java

```diff
@@ -637,15 +637,16 @@ public EnvelopePart slice(
         return sliced;
     }
 
-    /** Returns a new EnvelopePart, where all positions are shifted by positionDelta */
-    public EnvelopePart copyAndShift(double positionDelta) {
+    /** Returns a new EnvelopePart, where all positions are shifted by positionDelta.
+     * Resulting positions are limited at maxPosition. */
+    public EnvelopePart copyAndShift(double positionDelta, double maxPosition) {
         var newPositions = new DoubleArrayList();
         var newSpeeds = new DoubleArrayList();
         var newTimeDeltas = new DoubleArrayList();
         newPositions.add(positions[0] + positionDelta);
         newSpeeds.add(speeds[0]);
         for (int i = 1; i < positions.length; i++) {
-            var p = positions[i] + positionDelta;
+            var p = Math.min(maxPosition, positions[i] + positionDelta);
```

Comment:
are you sure we don't need a lower bound too? If so, can you `assert positionDelta >= 0`?

Reply:
I added a lower bound